### ISSUE TITLE
Postprocessing vocabulary cannot express flow-dynamics / NMS decoders used by modern instance-segmentation models

### ISSUE THREAD

**Issue by oeway** [bioimage-io/spec-bioimage-io#751]
## Summary

The current postprocessing vocabulary in v0.5 (`sigmoid`, `softmax`, `binarize`, `clip`, `scale_linear`, `scale_range`, `ensure_dtype`, …) can express the tail of a semantic-segmentation model but cannot express the tail of modern **instance-segmentation** models, where the network produces a dense auxiliary field that has to be decoded into per-object labels by an iterative / non-trivial algorithm that lives outside the neural network.

Concrete examples we'd like to publish on bioimage.io:

| model | network output | postprocessing needed |
| --- | --- | --- |
| **Cellpose / Cellpose-SAM** | `[flow_y, flow_x, cellprob]` | `cellpose.dynamics.compute_masks` — gradient tracking + connected components |
| **Omnipose** | euclidean flow + boundary | distance-transform-based watershed + Euler integration |
| **StarDist** | object probabilities + ray distances | NMS on rays → polygon/polyhedron rasterisation |
| **HoVer-Net** | horizontal + vertical gradient maps | marker-controlled watershed (custom) |

Today the only way to ship such a model on the zoo is the workaround I used for `Cellpose-SAM`: declare the *raw field* as the model output and tell users in the README to run the postprocessing step themselves with the vendor's Python package. That works, but:

1. The model card looks like a semantic-segmentation model, which is misleading; curators and users see a 3-channel logit tensor and have no standard way to turn it into a label map.
2. Zoo consumers (Ilastik, ImageJ, deepImageJ, the bioengine test runner) cannot execute the model end-to-end — they have no way to discover that a vendor-specific Python postprocessing step is required.
3. Every integration rewrites the same algorithm (there are at least three separate implementations of `compute_masks` in the wild already).

## Proposal — `custom` postprocessing op

Add a single new postprocessing descriptor modelled on `weights.pytorch_state_dict.architecture`:

```yaml
outputs:
  - id: masks
    axes: [...]
    postprocessing:
      - id: custom
        kwargs:
          source: flow_dynamics.py      # bundled in the package
          callable: compute_masks       # name of the Python callable
          sha256: ...                   # reproducibility hash
          environment: environment.yaml # reuse weights.*.dependencies semantics
          kwargs:                       # passed through to the callable
            cellprob_threshold: 0.0
```

Semantics:

- The callable receives the output tensor (`xarray.DataArray` or `np.ndarray`, matching the existing `*ProcessingDescr` conventions), returns a tensor with the declared output axes.
- Runs in an isolated conda/venv environment driven by the same `environment.yaml` / `dependencies` mechanism that `architecture.dependencies` already uses — no change to the JSON schema of dependencies, just reuse.
- Consumers that don't want to spawn a Python process (e.g. pure-Java deepImageJ) can still introspect the output tensor shape and flag the model as "needs vendor postprocessing".
- Marked clearly in the spec as a trust boundary — executing arbitrary vendor code. Curator review already gates the zoo, so this is acceptable; the alternative (everyone rolling their own sidecar registry) is worse.

This mirrors the shape of `architecture.source` / `architecture.callable`, which the community is already comfortable with.

## Alternative — a registry of named postprocessing algorithms

If executing arbitrary Python is too large a trust expansion, consider a registry of well-known non-trivial postprocessing ops:

```yaml
postprocessing:
  - id: cellpose_flow_dynamics
    kwargs:
      cellprob_threshold: 0.0
      flow_threshold: 0.4
```

…where the spec explicitly lists `cellpose_flow_dynamics`, `stardist_nms`, `hovernet_watershed`, etc., and each consumer SDK is expected to ship an implementation. This keeps the spec closed but grows linearly with the ecosystem.

## What I'd like feedback on

- Do the maintainers have a preferred direction between the two options above (or a third)?
- Is there already a WIP RFC or design note for this that I missed?
- Happy to prototype the `custom` descriptor against `bioimageio.spec` + `bioimageio.core` and open a PR if the direction is agreed.

## Context

Hit this while preparing a bioimage.io submission for Cellpose-SAM (the current Cellpose 4.x SAM-based model). The submission is now staged as a raw-flow model, but losing the postprocessing step in the spec feels like a gap worth fixing before more instance-seg models land.

**Comment by FynnBe** [comment]
stardist postprocessing is already defined in spec: https://bioimage-io.github.io/spec-bioimage-io/v0.5.9.1/api/bioimageio/spec/model/v0_5/?h=stardist#bioimageio.spec.model.v0_5.StardistPostprocessingDescr
and implemented in core: [2d](https://bioimage-io.github.io/spec-bioimage-io/v0.5.9.1/api/bioimageio/spec/model/v0_5/?h=stardist#bioimageio.spec.model.v0_5.StardistPostprocessingDescr), [3d](https://bioimage-io.github.io/core-bioimage-io-python/v0.10.2/api/bioimageio/core/proc_ops/?h=stardist#bioimageio.core.proc_ops.StardistPostprocessing3D) 

Arbitrary postprocessing can always be incorporated into the pytorch_state_dict format; the postprocessing field is designed to be a selected from a registry of known postprocessing operations.

Unlike stardist, cellpose is pytorch based, so we can choose the pytorch_state_dict route for it.
Adding a dedicated cellpose postprocessing op would only bring a benefit if it could be installed with no additional or reduced dependencies compared to the regular cellpose package. Otherwise cellpose models can be executed with the cellpose package available only anyway (albeit via the bioimageio.core api). Outside of Python there is little hope to run cellpose models via bioimage-io in the medium turn afaik.

**Comment by oeway** [comment]
This is what I think we can do now, model contributor can do:
```yaml
postprocessing:
  - id: my_custom_cellpose_flow_dynamics
    kwargs:
      cellprob_threshold: 0.0
      flow_threshold: 0.4
```
But before the model can be accepted, the need to submit a PR to our core repo to support `my_custom_cellpose_flow_dynamics`, which is a class or a function in Python, inside that function, they can add anything, and we can steer with the core package, like this we can at least allow extension of it.
Would you be ok with something like this?
@FynnBe

**Comment by FynnBe** [comment]
Could the actual cellpose transformer be saved as torchscript model with such a postprocessing?
Because if it still remains to be a custom pytorch_state_dict model there is no benefit to be gained from adding such a specific postprocessing.
I ask this as the [latest CellPose model implementation](https://github.com/MouseLand/cellpose/blob/a9f8bfcde43033247309e3982747df9fe9f09315/cellpose/models.py#L67) is not a valid `troch.nn.Module`. (and inference is not simply calling it's .net attribute (the transformer that it a torch.nn.Module).

Difference to the stardist use-case:
With stardist postprocessing now available in spec/core, we can now run inference via keras 3 with a pytorch backend on a (slightly modified/patched stardist tensorflow model), so this adds flexibility for deployment. (A light-weight stardist installation for postprocessing only is planned also).

If we cannot split cellpose into a 'pure' network + postprocessing (easily) or do not gain anything from it (can we convert the (not yet available) pure pytorch model to torchscript/onnx?), then there is no point in adding a custom postprocessing for cellpose instead of having cellpose as a custom model with pytorch_state_dict weights, which we can have with the spec and core as they are now.

**Comment by FynnBe** [comment]
As I have looked at the cellpose code before and decided to (at least as a first version) implement a wrapper to make it similar to a torch.nn.Module, I will likely add a protocol like this:
```python
@runtime_checkable
class PTorchModel(Protocol):
    @abstractmethod
    def load_state_dict(
        self, state_dict: Mapping[str, Any], strict: bool = True, assign: bool = False
    ) -> Self: ...

    @abstractmethod
    def to(self, ) -> Self: ...

    @abstractmethod
    def forward(self, *input: torch.Tensor) -> Sequence[torch.Tensor]: ...

    def eval(self) -> Self:
        """Set model to eval mode"""
        return self
```
to bioimageio.core.backends.pytorch_backend and make sure that the object returned by `callable` for pytorch_state_dict weights only has to comply with this protocol and not be a valid torch.nn.Module.

I hopw this will help other similar cases to avoid having to rewrite their model as a valid torch.nn.Module.

**Comment by FynnBe** [comment]
(taking naming suggestions to avoid the error-prong `PTorchModel` (vs `PyTorchModule`...) @k-dominik )

**Comment by k-dominik** [comment]
Since it's emulating parts of Module... `TorchNNModule(Protocol)`, `TorchNNModuleMinimal(Protocol)`, or `TorchBioimageIOModule(Protocol)`

and for fun, of course `PsyTorchModel` :P

**Comment by FynnBe** [comment]
landed on `TorchNNModuleLike` for now , thx @k-dominik

**Comment by oeway** [comment]
The point is to have a universal spec, simply anything take in numpy-like and spit out numpy-like  can work, CellPose is not the point here, we can open it up for anyone to add any pre/postprocessing, once they can make it work in our code base. One could run a imagej or rust based pre/post processing once they produce something, we just need to make the whole thing work. Then we will move the discussion on each individual case on what's the best way to add the specific operator into our repo, either port it to numpy or pytorch or keep it in imagej, that's we can decide there. But the spec should not try to prevent people from adding new processor -- otherwise, people will just go away. Nothing stops us from first take in a messy ugly implementation of a new processor and we revise later in future releases. But it has a real value to things move on, get people attached to the model zoo first.

**Comment by FynnBe** [comment]
How about we promote the use of pytorxh state dict format for this purpose. That has the huge advantage that models are usuable right away. It might also be possible in some cases to script or trace through that postprocessing leading to torchscript/onnx weights that include the postprocessing. 

Another problem I see with encouraging custom postprocessing is that we either accept them all which could lead to an unmaintainable mess or reject some which will make them turn away from the zoo and leaves an unusable model...

If we have real use cases for custom ops outside pytorch I would rather add a 'custom' weights field and treat it analog to pytorch state dict (just not fixed to torch backend).
This would at least avoid having to merge two or more sets (weights and postprocessing) of additional dependencies.

**Comment by oeway** [comment]
Not sure if we can, the whole point of why people have to split the model into pre/postprocessing is that most of the operations cannot be expressed using neural framework torchscript/onnx, a postprocessing might do a watershed, or have if/else loop, using numpy/numba etc. so it has to be universal operators.

we are not encouraging custom postprocessing, we are providing the vocabulary to express the model user already developed and decided on the postprocessing they want, either they find in our list they can proceed, or they cannot and they walk away.  Not sure if we have the choice.

It's not realistic to ask all the contributor to rewrite postprocessing in pytorch, it will be hard, or impossible, can you write a watershed algorithm in pytorch?  and distance transformation, connected components? I don't see the point, i would rather just use scikit-image. And these should be global, not tied to pytorch at all, it should work for any model for segmentation, look at a real example here for useful postprocessing: example_descriptions/models/unet2d_keras_tf2/Contours2InstanceSegmentation.ijm

and imagej had to extend its own preprocessing: https://github.com/bioimage-io/spec-bioimage-io/blob/9893003752b4768467bf693f73119be519944050/example_descriptions/models/unet2d_keras_tf2/v0_4.bioimageio.yaml#L37-L38

IMO, forcing pre/postprocessing only pytorch, and only support in pytorch state dict is the wrong appraoch, it cannot solve most of the problem, and if I can write them in pytorch, then I won't even need pre/postprocessing, i could simply implement then in my architcture python file, no?

**Comment by oeway** [comment]
@akreshuk — opening PR #752 to move this forward.

**Design decision:** We went with a single flat `CustomPostprocessingDescr` (id: `"custom"`) rather than a more elaborate hierarchy. The reason: the overwhelming feedback in this thread is that the contributor interface must be as simple as possible.

**Minimal API — four fields:**

| Field | Required | Purpose |
|---|---|---|
| `callable` | yes | function or class name |
| `source` | dev only | inline `.py` file shipped with the package |
| `sha256` | when `source` given | integrity check |
| `kwargs` | no | forwarded to `__init__` / factory |

**Inline → built-in promotion is transparent:** `callable` and `kwargs` are identical in both stages; only `source` + `sha256` are dropped once the op is merged into `custom_ops/`.

We also added `custom_ops/cellpose_flow_dynamics.py` as the first built-in op and a full contributor README (`custom_ops/README.md`) showing both callable-class and factory-function styles.

Happy to adjust scope or naming before merge — please review at #752.

**Comment by FynnBe** [comment]
To show you that we would not need custom ops for cellpose, I'll drop a draft of the naive CellposeWrapper I wrote here that can be used in the pytorch_state_dict weights format:
```python
from tempfile import NamedTemporaryFile
from typing import Any, Mapping, Optional

import torch
from cellpose import models
from loguru import logger
from typing_extensions import Self

from bioimageio.core.backends.pytorch_backend import TorchNNModuleLike


class CellposeWrapper(TorchNNModuleLike):
    def __init__(self, *, use_bfloat16: bool = True):
        super().__init__()
        self.use_bfloat16 = use_bfloat16
        self._model = None

    def load_state_dict(
        self, state_dict: Mapping[str, Any], strict: bool = True, assign: bool = False
    ) -> Self:
        with NamedTemporaryFile(suffix=".cpsam") as tmp:
            torch.save(state_dict, tmp.name)
            self._model = models.CellposeModel(
                gpu=False, pretrained_model=tmp.name, use_bfloat16=self.use_bfloat16
            )

        return self

    def to(
        self,
        *,
        device: Optional[torch.device] = None,
        dtype: Optional[torch.dtype] = None,
        non_blocking: bool = False,
    ) -> Self:
        if self._model is None:
            raise ValueError("Model not loaded yet")

        if dtype is not None:
            logger.warning(
                "possibly ignoring `use_bfloat16` argument by setting dtype explicitly"
            )
            self._model.net = self._model.net.to(dtype=dtype)

        if device is not None:
            self.device = device
            self._model.net = self._model.net.to(device)

        return self

    def forward(self, *input: torch.Tensor) -> torch.Tensor:
        if self._model is None:
            raise ValueError("Model not loaded yet")
        if len(input) != 1:
            raise ValueError("Expected exactly one input tensor")

        return self._model.eval(input[0])[0]  # pyright: ignore[reportUnknownVariableType]
```

TorchNNModuleLike protocol is [here](https://github.com/bioimage-io/core-bioimage-io-python/blob/257ea240272b2e9cd8b76d91b9114109db4965f4/src/bioimageio/core/backends/pytorch_backend.py#L25-L48)

**Comment by FynnBe** [comment]
I do agree that it is cleaner to have postprocessing separated though, so if you want to go ahead with that I have some suggestions:

1. Is it not foreseeable that (custom) postprocessing may act on more than one output?
   Maybe time to go ahead with moving pre and postprocessing out of `inputs`/`outputs` https://github.com/bioimage-io/spec-bioimage-io/issues/524. This would also promote its significance.
2. If we incorporate new ops into spec/core they are no longer 'custom', so I would just call them 'ops' (rather than the current submodule name 'proc_ops' maybe). Of course as long as there is `source` and `kwargs`, the (implicit) id is 'custom'.
3. I would enforce to inherit from one of three base classes: https://github.com/bioimage-io/core-bioimage-io-python/blob/4fadbc158efbe853158d52022730130a83e33d67/src/bioimageio/core/_op_base.py#L32-L44
This keeps the same interface for ops acting on multiple tensors and makes it transparent if the op can be used blockwise or not and gives access to dataset/sample statistics via the sample object.
4. To deal with dependencies of custom ops we would need optional `dependencies` as conda dependencies. I would constrain that all dependencies within a model share the same conda channels and have identical pins if they list the same dependency. This would allow us to merge dependencies unambiguously.
5. Op implementations should go into bioimageio.core

**Comment by FynnBe** [comment]
When we agree on the design decisions I can also implement the rest.

**Comment by oeway** [comment]
> TorchNNModuleLike

This looks like a hack to me -- `TorchNNModuleLike ` is not for this, TorchNNModule should be able to support backprop, and postprocessing are not necessary differentiable. And this will only works for pytorch state dict, it's like to say if you want to use custom ops, you have to use pytorch, and no torch scripts, onnx and so on. I mean if the point is to get the model run, we could remove the entire spec, and ask user to submit a python file with a class that is TorchNNModuleLike, and some light weight authors and so on.

> Is it not foreseeable that (custom) postprocessing may act on more than one output?

That sound like a breaking change, I'd do one thing at a time, the custom ops i propose here won't break anything, if you want to do that, please do it in another PR or version, and ensure backward compatibility somehow, or do a batch processing script to upgrade all the models in the model zoo.

> If we incorporate new ops into spec/core they are no longer 'custom', so I would just call them 'ops' (rather than the current submodule name 'proc_ops' maybe). Of course as long as there is source and kwargs, the (implicit) id is 'custom'.
> I would enforce to inherit from one of three base classes: https://github.com/bioimage-io/core-bioimage-io-python/blob/4fadbc158efbe853158d52022730130a83e33d67/src/bioimageio/core/_op_base.py#L32-L44
This keeps the same interface for ops acting on multiple tensors and makes it transparent if the op can be used blockwise or not and gives access to dataset/sample statistics via the sample object.

That in principle make sense, however, I think that's too complicated for most people, they won't have time to really figure out what you mean about required_measures, blockwise or not, ....  (think attention span, after some minutes of browsing, if they cannot understand and how to write the base class with required_measures or so, they/I will go away-- i browsed it briefly and I canot figure it out what you mean by required_measures), and I double that most use case need it, the cases we have in our hand, cellpose and stardist, and some other simple thing, we don't need to care about any of these, I'd rather do a very simple function/class interface -- with that said, at anytime, if we don't like it, we can take over, and reimplement it with the blockwise and so on, but forcing people to understand our internal interface, seems to be overkill for most cases and virtual block many contributors.

> To deal with dependencies of custom ops we would need optional dependencies as conda dependencies. I would constrain that all dependencies within a model share the same conda channels and have identical pins if they list the same dependency. This would allow us to merge dependencies unambiguously.

in principle, yes, however, I think we don't need it now, given that most postprocessing can handle by numpy etc. standard library, we can extend it later when we see it, and nothing breaks. 


> Op implementations should go into bioimageio.core

Make sense!



I would separate the concern and try to do a lot of things in one PR, take it easy and one thing a time, especially the ones that won't be breaking anything, but can have big impact immediately. There are always other possibilities, that extensions, and I would do it later when we see that it's absolutely necessary.
